File: mobile/android/app/src/test/java/app/terra/explore/ActivityMetricsTests.kt
package app.terra.explore
import org.junit.Test
import org.junit.Assert.*
class ActivityMetricsTests {
    private fun walk()=Session(mode=Mode.walk,startedAt=1000,points=mutableListOf(Point(0.0,0.0,1000,5.0),Point(0.0001,0.0,11000,5.0)))
    @Test fun speedExpiresAndPauses() {
        assertEquals(4.003,ActivityMetrics.speed(walk(),11000),0.02)
        assertEquals(0.0,ActivityMetrics.speed(walk(),30000),0.0)
        assertEquals(0.0,ActivityMetrics.speed(walk().copy(pausedAt=11000),11000),0.0)
    }
    @Test fun energyUsesWeightAndFallback() {
        val base=ActivityMetrics.calories(walk(),70.0)
        assertTrue(base>0); assertEquals(base*2,ActivityMetrics.calories(walk(),140.0),0.001)
        assertEquals(base,ActivityMetrics.calories(walk(),null),0.0)
        assertEquals(base,ActivityMetrics.calories(walk(),Double.NaN),0.0)
    }
    @Test fun noEnergyAcrossGapsOrInCar() {
        val s=walk(); s.points[1]=s.points[1].copy(gap=true); assertEquals(0.0,ActivityMetrics.calories(s,null),0.0)
        s.points[1]=s.points[1].copy(t=90000,gap=false); assertEquals(0.0,ActivityMetrics.calories(s,null),0.0)
        assertEquals(0.0,ActivityMetrics.calories(walk().copy(mode=Mode.car),null),0.0)
    }
}

File: mobile/ios/TerraTests/ActivityMetricsTests.swift
import XCTest
@testable import Terra
final class ActivityMetricsTests: XCTestCase {
    private var walk: TrackSession { TrackSession(mode: .walk,startedAt: 1000,points: [TrackPoint(lat: 0,lng: 0,t: 1000,accuracy: 5),TrackPoint(lat: 0.0001,lng: 0,t: 11000,accuracy: 5)]) }
    func testFreshSpeedAndPause() {
        XCTAssertEqual(ActivityMetrics.speed(walk,now: 11000),4.003,accuracy: 0.02)
        XCTAssertEqual(ActivityMetrics.speed(walk,now: 30000),0)
        var paused = walk; paused.pausedAt = 11000; XCTAssertEqual(ActivityMetrics.speed(paused,now: 11000),0)
    }
    func testEnergyWeightAndFallback() {
        let base = ActivityMetrics.calories(walk,weight: 70)
        XCTAssertGreaterThan(base,0); XCTAssertEqual(ActivityMetrics.calories(walk,weight: 140),2*base,accuracy: 0.001)
        XCTAssertEqual(ActivityMetrics.calories(walk,weight: nil),base)
        XCTAssertEqual(ActivityMetrics.calories(walk,weight: .nan),base)
    }
    func testNoEnergyAcrossPauseOrInCar() {
        var s = walk; s.points[1].breakBefore = true; XCTAssertEqual(ActivityMetrics.calories(s,weight: nil),0)
        s = walk; s.points[1].t = 90000; XCTAssertEqual(ActivityMetrics.calories(s,weight: nil),0)
        s = walk; s.mode = .car; XCTAssertEqual(ActivityMetrics.calories(s,weight: nil),0)
    }
}
